### PR TITLE
FEATURE: update mermaid and add auto sizing

### PR DESCRIPTION
**Pull request by SamSaffron**
- Update mermaid to version 9.2.2
- Deliver asset locally instead of relying on third party CDN
- Automatically set height for mermaid element

### CODE REVIEW

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js [hunk 1]
```diff
@@ -58,9 +59,55 @@ async function applyMermaid(element, key = "composer") {
...
+  let height = parseInt(mermaid.getBoundingClientRect().height);
+  let calculatedHeight = parseInt(mermaid.dataset.calculatedHeight);
+
+  if (height === 0) {
+    return;
+  }
+
+  if (height !== calculatedHeight) {
+    mermaid.dataset.calculatedHeight = height;
+    // TODO: need to use API here
```
Comment (martin-brennan) [inline javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:79]: Which API are you meaning?

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js [hunk 2]
```diff
@@ -58,9 +59,55 @@ async function applyMermaid(element, key = "composer") {
...
+    let composer = document.getElementsByClassName("d-editor-input")[0];
```
Comment (martin-brennan) [inline javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:80]: Can just do `document.querySelector(".d-editor-input")`, that will return the first matching element of that class

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js [hunk 3]
```diff
@@ -58,9 +59,55 @@ async function applyMermaid(element, key = "composer") {
...
+    let old = composer.value;
```
Comment (martin-brennan) [inline javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:81]: Not sure why `old` is needed just for the split? Did you mean to use it later down in `(joined !== composer.value)`? If so `oldComoserValue` would be more descriptive, if not IMO just remove it and use `composer.value`

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js [hunk 4]
```diff
@@ -58,9 +59,55 @@ async function applyMermaid(element, key = "composer") {
...
+
+    let split = old.split("\n");
+
+    let n = 0;
+    for (let i = 0; i < split.length; i++) {
+      if (split[i].match(/```mermaid/)) {
+        if (n === index) {
+          split[i] = "```mermaid height=" + height;
```
Comment (martin-brennan) [inline javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:89]: Are there other options for mermaid that could be present here that we are wiping out? Or is `height` the only one?

**Review by martin-brennan — COMMENTED** [review]
I noticed in the composer if I remove the auto-height or even change it we re-insert it straight away. This may be a little confusing for users? Maybe we should just not do anything if the `height` is already part of the `\```mermaid` line? Other comments are mainly about the JS code.

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js (hunk 4 above)
Comment (SamSaffron) [reply javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:89]: height is all we got as far as I know.

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js (hunk 2 above)
Comment (SamSaffron) [reply javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:80]: oh ... I mean some sort of API to hunt for the composer, vs getElementsByClassName

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js (hunk 3 above)
Comment (SamSaffron) [reply javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:81]: true, yeah will remove

**Comment by SamSaffron** [comment]
> I noticed in the composer if I remove the auto-height or even change it we re-insert it straight away. This may be a little confusing for users?  

Super tricky cause as you our editing the height can be changing, you don't want the height to become sticky when the diagram is tiny. You want stuff to react to edits. Maybe we add an attribute to denote it is an auto height? 

eg:

````
```mermaid height=200,auto
````

Will give it a shot.

**Comment by SamSaffron** [comment]
@martin-brennan all should be addressed now

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js [hunk 5]
```diff
@@ -58,9 +59,55 @@ async function applyMermaid(element, key = "composer") {
       mermaid.dataset.processed = true;
       mermaid.querySelector(".spinner")?.remove();
     }
+
+    if (key === "composer") {
+      debounce(() => updateMarkdownHeight(mermaid, index), 500);
```
Comment (jjaffeux) [inline javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:64]: That debounce is useless. If you add a `console.log` in `updateMarkdownHeight` you will see it's called constantly while typing in composer.

The reason for this is that debounce can't work with an anonymous function, as it recreates a new context each time, I can fix it for you if you want, let me know.

**Comment by jjaffeux** [comment]
Other than my comment, it seems to work very nicely ❤️

**Review by martin-brennan — APPROVED** [review]
auto height change makes sense, and thanks for doing the other fixes too :)

File: javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js (hunk 5 above)
Comment (SamSaffron) [reply javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:64]: yeah ... half useless, at least it delays the call... deanonymising this function is a pita :( can you fix for me and merge, would love to see the fix.

Comment (jjaffeux) [reply javascripts/discourse/api-initializers/discourse-mermaid-theme-component.js:64]: Will look today

**Comment by jjaffeux** [comment]
@SamSaffron merged 🎉

**Pull request merged**